File: frontend/src/api/studentAnalytics.ts
import { fetchApi } from '../lib/api';

export interface AnalyticsOverview {
  study_hours_week: number;
  study_hours_month: number;
  attendance_percentage: number;
  quiz_average: number;
  assignments_completed: number;
  assignments_total: number;
  interview_score: number;
  documents_uploaded: number;
  question_papers_solved: number;
  placement_readiness: number;
  weekly_progress: number;
  monthly_progress: number;
}

export interface AiInsightItem {
  type: 'warning' | 'tip' | 'success' | 'info';
  title: string;
  message: string;
  subject?: string;
  action_label?: string;
  action_url?: string;
}

export interface AiInsightsResponse {
  insights: AiInsightItem[];
  generated_at: string;
}

export interface ChartDataPoint {
  label: string;
  value: number;
  secondary_value?: number;
}

export interface SubjectPerformancePoint {
  subject: string;
  quiz_score: number;
  attendance: number;
  assignments: number;
}

export interface AnalyticsChartsData {
  weekly_study_hours: ChartDataPoint[];
  monthly_study_hours: ChartDataPoint[];
  quiz_performance: ChartDataPoint[];
  attendance_trend: ChartDataPoint[];
  assignment_completion: ChartDataPoint[];
  interview_scores: ChartDataPoint[];
  subject_progress: SubjectPerformancePoint[];
}

export interface StreakData {
  daily_study_streak: number;
  quiz_streak: number;
  attendance_streak: number;
  assignment_streak: number;
  last_activity_date?: string;
}

export interface GoalResponse {
  id: number;
  user_id: number;
  title: string;
  category: string;
  target_metric: string;
  target_value: number;
  current_value: number;
  unit: string;
  deadline?: string;
  status: string;
  progress_percentage: number;
  created_at: string;
}

export interface GoalCreatePayload {
  title: string;
  category: string;
  target_metric: string;
  target_value: number;
  current_value: number;
  unit: string;
  deadline?: string;
}

export interface GoalUpdatePayload {
  title?: string;
  current_value?: number;
  target_value?: number;
  status?: string;
  deadline?: string;
}

export interface AnalyticsExportPayload {
  student_name: string;
  export_date: string;
  overview: AnalyticsOverview;
  charts: AnalyticsChartsData;
  goals: GoalResponse[];
  streaks: StreakData;
  insights: AiInsightItem[];
}

export const getAnalyticsOverview = async (): Promise<AnalyticsOverview> => {
  return fetchApi('/analytics/overview');
};

export const getAiInsights = async (): Promise<AiInsightsResponse> => {
  return fetchApi('/analytics/ai-insights');
};

export const getAnalyticsCharts = async (): Promise<AnalyticsChartsData> => {
  return fetchApi('/analytics/charts');
};

export const getStreaks = async (): Promise<StreakData> => {
  return fetchApi('/analytics/streaks');
};

export const getGoals = async (): Promise<GoalResponse[]> => {
  return fetchApi('/analytics/goals');
};

export const createGoal = async (payload: GoalCreatePayload): Promise<GoalResponse> => {
  return fetchApi('/analytics/goals', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
};

export const updateGoal = async (id: number, payload: GoalUpdatePayload): Promise<GoalResponse> => {
  return fetchApi(`/analytics/goals/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
};

export const deleteGoal = async (id: number): Promise<void> => {
  return fetchApi(`/analytics/goals/${id}`, {
    method: 'DELETE',
  });
};

export const exportAnalytics = async (): Promise<AnalyticsExportPayload> => {
  return fetchApi('/analytics/export');
};

File: backend/app/schemas/student_analytics.py
from typing import Optional, List
from datetime import date, datetime
from pydantic import BaseModel


# ── Overview Metrics ──────────────────────────────────────────────────────────

class AnalyticsOverview(BaseModel):
    study_hours_week: float
    study_hours_month: float
    attendance_percentage: float
    quiz_average: float
    assignments_completed: int
    assignments_total: int
    interview_score: float
    documents_uploaded: int
    question_papers_solved: int
    placement_readiness: float
    weekly_progress: float
    monthly_progress: float


# ── AI Insights ───────────────────────────────────────────────────────────────

class AiInsightItem(BaseModel):
    type: str  # warning, tip, success, info
    title: str
    message: str
    subject: Optional[str] = None
    action_label: Optional[str] = None
    action_url: Optional[str] = None


class AiInsightsResponse(BaseModel):
    insights: List[AiInsightItem]
    generated_at: datetime


# ── Chart Data ────────────────────────────────────────────────────────────────

class ChartDataPoint(BaseModel):
    label: str
    value: float
    secondary_value: Optional[float] = None


class SubjectPerformancePoint(BaseModel):
    subject: str
    quiz_score: float
    attendance: float
    assignments: float


class AnalyticsChartsData(BaseModel):
    weekly_study_hours: List[ChartDataPoint]
    monthly_study_hours: List[ChartDataPoint]
    quiz_performance: List[ChartDataPoint]
    attendance_trend: List[ChartDataPoint]
    assignment_completion: List[ChartDataPoint]
    interview_scores: List[ChartDataPoint]
    subject_progress: List[SubjectPerformancePoint]


# ── Streaks ───────────────────────────────────────────────────────────────────

class StreakResponse(BaseModel):
    daily_study_streak: int
    quiz_streak: int
    attendance_streak: int
    assignment_streak: int
    last_activity_date: Optional[date] = None


# ── Goals ─────────────────────────────────────────────────────────────────────

class GoalCreateRequest(BaseModel):
    title: str
    category: str = "Academic"
    target_metric: str
    target_value: float
    current_value: float = 0.0
    unit: str = "%"
    deadline: Optional[date] = None


class GoalUpdateRequest(BaseModel):
    title: Optional[str] = None
    current_value: Optional[float] = None
    target_value: Optional[float] = None
    status: Optional[str] = None
    deadline: Optional[date] = None


class GoalResponse(BaseModel):
    id: int
    user_id: int
    title: str
    category: str
    target_metric: str
    target_value: float
    current_value: float
    unit: str
    deadline: Optional[date] = None
    status: str
    progress_percentage: float
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Export ────────────────────────────────────────────────────────────────────

class AnalyticsExportPayload(BaseModel):
    student_name: str
    export_date: datetime
    overview: AnalyticsOverview
    charts: AnalyticsChartsData
    goals: List[GoalResponse]
    streaks: StreakResponse
    insights: List[AiInsightItem]
